File: src/main/java/net/dean/jraw/http/oauth/Credentials.java
package net.dean.jraw.http.oauth;

import net.dean.jraw.http.AuthenticationMethod;
import net.dean.jraw.util.JrawUtils;

import java.net.URI;
import java.util.UUID;

/**
 * This class is responsible for handling any kind of Reddit user and OAuth2 credentials
 */
public final class Credentials {
    private final AuthenticationMethod authenticationMethod;
    private final String username;
    private final String password;
    private final String clientId;
    private final String clientSecret;
    private final UUID deviceId;
    private final URI redirectUri;

    private Credentials(AuthenticationMethod authenticationMethod, String username, String password, String clientId,
                        String clientSecret, String redirectUri) {
        this(authenticationMethod, username, password, clientId, clientSecret, null, redirectUri);
    }

    public Credentials(AuthenticationMethod authenticationMethod, String username, String password, String clientId,
                       String clientSecret, UUID deviceId, String redirectUri) {
        this.authenticationMethod = authenticationMethod;
        this.username = username;
        this.password = password;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.deviceId = deviceId;
        this.redirectUri = redirectUri != null ? JrawUtils.newUri(redirectUri) : null;
    }

    /**
     * Gets how these Credentials are to be used
     * @return The AuthenticationMethod
     */
    public AuthenticationMethod getAuthenticationMethod() {
        return authenticationMethod;
    }

    public String getUsername() {
        if (username == null)
            throw new IllegalStateException("This method is not appropriate for this authentication method");
        return username;
    }

    public String getPassword() {
        if (username == null)
            throw new IllegalStateException("This method is not appropriate for this authentication method");
        return password;
    }

    /**
     * Gets the OAuth2 app's ID. Applicable for all OAuth app types.
     * @return The app's ID
     */
    public String getClientId() {
        return clientId;
    }

    /**
     * Gets the OAuth2 app's secret key. Applicable for all OAuth app types, however, for 'installed app' app types, the
     * secret will always be an empty string. All parameters must be non-null.
     *
     * @return The secret value for the application
     */
    public String getClientSecret() {
        return clientSecret;
    }

    /**
     * Gets a UUID that is unique for every device. Only non-null when instantiated using
     * {@link #userlessApp(String, UUID)}.
     */
    public UUID getDeviceId() {
        return deviceId;
    }

    public URI getRedirectUri() {
        return redirectUri;
    }

    private static void assertNotNull(Object... objects) {
        for (Object o : objects)
            if (o == null)
                throw new NullPointerException("Arguments must not be null");
    }

    /**
     * Creates a new Credentials object for a Reddit app whose type is 'installed app.' All parameters must be non-null.
     * @param clientId The publicly-available app ID
     * @return A new Credentials
     */
    public static Credentials installedApp(String clientId, String redirectUri) {
        // Use an empty string as the client secret because "The 'password' for non-confidential clients
        // (installed apps) is an empty string". See https://github.com/reddit/reddit/wiki/OAuth2#token-retrieval
        assertNotNull(clientId, redirectUri);
        return new Credentials(AuthenticationMethod.APP, null, null, clientId, "", redirectUri);
    }

    /**
     * Creates a new Credentials object for a Reddit app whose type is 'web app.' All parameters must be non-null.
     *
     * @param username The app owner's username
     * @param password The app owner's password
     * @param clientId The publicly-available app ID
     * @param clientSecret The secret value for the application
     * @return A new Credentials
     */
    public static Credentials script(String username, String password, String clientId, String clientSecret) {
        assertNotNull(username, password, clientId, clientSecret);
        return new Credentials(AuthenticationMethod.SCRIPT, username, password, clientId, clientSecret, null);
    }

    /**
     * Creates a new Credentials object for a Reddit app whose type is 'web app.' All parameters must be non-null.
     *
     * @param username The app owner's username
     * @param password The app owner's password
     * @param clientId The publicly-available app ID
     * @param clientSecret The secret value for the application
     * @param redirectUri The uri to be redirected to
     * @return A new Credentials
     */
    public static Credentials script(String username, String password, String clientId, String clientSecret, String redirectUri) {
        assertNotNull(username, password, clientId, clientSecret);
        return new Credentials(AuthenticationMethod.SCRIPT, username, password, clientId, clientSecret, redirectUri);
    }

    /**
     * Creates a new Credentials object for a Reddit app whose type is 'web app.' All parameters must be non-null.
     * @param clientId The publicly-available app ID
     * @param clientSecret The secret value for the application
     * @return A new Credentials
     */
    public static Credentials webapp(String clientId, String clientSecret, String redirectUri) {
        assertNotNull(clientId, clientSecret, redirectUri);
        return new Credentials(AuthenticationMethod.WEBAPP, null, null, clientId, clientSecret, redirectUri);
    }

    /**
     * Creates a new Credentials object for using the API in a user-less context. All parameters must be non-null
     *
     * @param deviceId A unique, per-device ID.
     * @see AuthenticationMethod#USERLESS
     */
    public static Credentials userless(String clientId, String clientSecret, UUID deviceId) {
        assertNotNull(clientId, clientSecret, deviceId);
        return new Credentials(AuthenticationMethod.USERLESS, null, null, clientId, clientSecret, deviceId, null);
    }

    /**
     * Creates a new Credentials object for using the API in a user-less context on an installed application. All
     * parameters must be non-null.
     *
     * @param deviceId The UUID unique to this device
     * @see AuthenticationMethod#USERLESS_APP
     */
    public static Credentials userlessApp(String clientId, UUID deviceId) {
        // Use an empty string as the client secret because "The 'password' for non-confidential clients
        // (installed apps) is an empty string". See https://github.com/reddit/reddit/wiki/OAuth2#token-retrieval
        return new Credentials(AuthenticationMethod.USERLESS_APP, null, null, clientId, "", deviceId, null);
    }
}

File: lib/src/main/kotlin/net/dean/jraw/oauth/Credentials.kt
package net.dean.jraw.oauth

import java.util.*

/** This is a data class that represents the credentials to an OAuth2 app */
class Credentials private constructor(
    /** How these credential were meant to be used */
    val authMethod: AuthMethod,

    /** Reddit username. Only non-null for script Credentials */
    val username: String? = null,

    /** Reddit password. Only non-null for script Credentials */
    val password: String? = null,

    /** OAuth2 app client ID. */
    val clientId: String,

    /** OAuth2 app client secret. This is an empty string for app or userless app Credentials. */
    val clientSecret: String,

    /** A unique, per-device ID */
    val deviceId: UUID? = null,

    /**
     * The URL that users will be redirected to when authorizing your application. Only non-null for installed or web
     * apps.
     */
    val redirectUrl: String? = null
) {
    /** Factory construction methods */
    companion object {
        /** Creates credentials for a script app */
        @JvmStatic fun script(username: String, password: String, clientId: String, clientSecret: String) =
            Credentials(AuthMethod.SCRIPT,
                username = username,
                password = password,
                clientId = clientId,
                clientSecret = clientSecret)

        /** Creates credentials for an installed app (Android, iOS, etc.) */
        @JvmStatic fun installedApp(clientId: String, redirectUrl: String) =
            Credentials(AuthMethod.APP,
                clientId = clientId,
                clientSecret = "",
                redirectUrl = redirectUrl)

        /** Creates credentials for a web app */
        @JvmStatic fun webapp(clientId: String, clientSecret: String, redirectUrl: String) =
            Credentials(AuthMethod.WEBAPP,
                clientId = clientId,
                clientSecret = clientSecret,
                redirectUrl = redirectUrl)

        /** Creates credentials for a script or web app running without the context of a user */
        @JvmStatic fun userless(clientId: String, clientSecret: String, deviceId: UUID) =
            Credentials(AuthMethod.USERLESS,
                clientId = clientId,
                clientSecret = clientSecret,
                deviceId = deviceId)

        /** Creates credentials for an installed app running without the context of a user */
        @JvmStatic fun userlessApp(clientId: String, deviceId: UUID) =
            Credentials(AuthMethod.USERLESS_APP,
                clientId = clientId,
                clientSecret = "",
                deviceId = deviceId)
    }
}
